### PR TITLE
Fix no empty input check in newSignatureFromCommitline for gogit version

### LINKED ISSUE #28840: slice bounds out of range in  function: /modules/git.newSignatureFromCommitline
### Description

create a mirror from linux kernel repo, it clone the repo to disk successful but prompt "panic: runtime error: slice bounds out of range [:-1]" on the web. then i use cmd: gitea.exe admin repo-sync-releases, it show below log:

2024/01/18 15:02:39 .../setting/security.go:168:loadSecurityFrom() [W] Enabling Query API Auth tokens is not recommended. DISABLE_QUERY_AUTH_TOKEN will default to true in gitea 1.23 and will be removed in gitea 1.24.
panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
code.gitea.io/gitea/modules/git.newSignatureFromCommitline({0xc001ebacb8, 0x0, 0x20})
        /source/modules/git/signature_gogit.go:37 +0x3fa
code.gitea.io/gitea/modules/git.parseTagRef(0xc001ebafd8?)
        /source/modules/git/repo_tag.go:184 +0x22f
code.gitea.io/gitea/modules/git.(*Repository).GetTagInfos(0xc001cba1e0, 0x0, 0x3a071c5?)
        /source/modules/git/repo_tag.go:142 +0x7fc
code.gitea.io/gitea/modules/repository.pullMirrorReleaseSync({0x4b4c9f0, 0xc000d4bcc0}, 0xc0020c6800, 0x27?)
        /source/modules/repository/repo.go:502 +0x114
code.gitea.io/gitea/modules/repository.SyncReleasesWithTags({0x4b4c9f0?, 0xc000d4bcc0}, 0xc0020c6800, 0xc001cba1e0)
        /source/modules/repository/repo.go:294 +0x18a
code.gitea.io/gitea/cmd.runRepoSyncReleases(0xc00045d720?)
        /source/cmd/admin.go:392 +0x5ca
code.gitea.io/gitea/cmd.prepareSubcommandWithConfig.prepareWorkPathAndCustomConf.func1(0x2f16c00?)
        /source/cmd/main.go:117 +0x1ec
github.com/urfave/cli/v2.(*Command).Run(0x2f16c00, 0xc000ebd380, {0xc0001260e0, 0x1, 0x1})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:274 +0x998
github.com/urfave/cli/v2.(*Command).Run(0x2f18d00, 0xc000ebd080, {0xc0001eb200, 0x2, 0x2})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:267 +0xbe5
github.com/urfave/cli/v2.(*Command).Run(0xc001bf78c0, 0xc000ebcdc0, {0xc000194140, 0x3, 0x4})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:267 +0xbe5
github.com/urfave/cli/v2.(*App).RunContext(0xc0009683c0, {0x4b4c780?, 0x6960640}, {0xc000194140, 0x3, 0x4})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/app.go:332 +0x5b7
github.com/urfave/cli/v2.(*App).Run(...)
        /go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/app.go:309
code.gitea.io/gitea/cmd.RunMainApp(0xc0009683c0, {0xc000194140?, 0xc0020530e0?, 0x4a?})
        /source/cmd/main.go:176 +0x39
main.main()
        /source/main.go:46 +0x79

### Gitea Version

1.21.4

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

v1.21.4 is downloaded from github release，and used in win11x64

### Database

SQLite

### PR DESCRIPTION
**Pull request by yp05327**
Fix #28840

In signature_nogogit,go:
image

In signature_gogit.go:
image

### CODE REVIEW

**Comment by lunny** [comment]
Can you add some tests?

**Comment by yp05327** [comment]
How to add test in both nogit and gogit version? 🤔

**Comment by lunny** [comment]
Have you tested Linux kernal?

**Comment by lunny** [comment]
> How to add test in both nogit and gogit version? 🤔

It seems you have to changed twice at the moment.

**Comment by yp05327** [comment]
> Have you tested Linux kernal?

Yes. It works well.

**Comment by yp05327** [comment]
> It seems you have to change them twice at the moment.

A copy from `signature_gogit_test.go` is allowed? But seems no better method.

**Comment by lunny** [comment]
> > It seems you have to change them twice at the moment.
> 
> A copy from `signature_gogit_test.go` is allowed? But seems no better method.

You can copy first and then try to extract a function

**Comment by yp05327** [comment]
I understanded now. 
Just remove `//go:build gogit` in the `test.go`, then we can test it in both gogit version and no gogit version. 😄

File: modules/git/signature_gogit.go [hunk 1]
```diff
@@ -21,38 +21,59 @@ type Signature = object.Signature
 // Helper to get a signature from the commit line, which looks like these:
 //
 //	author Patrick Gundlach <gundlach@speedata.de> 1378823654 +0200
-//	author Patrick Gundlach <gundlach@speedata.de> Thu, 07 Apr 2005 22:13:13 +0200
+//	author Patrick Gundlach <gundlach@speedata.de> Thu Apr 07 22:13:13 2005 +0200
```
Comment (wxiaoguang) [inline modules/git/signature_gogit.go:24]: Why the code is different from nogogit? And it seems that the nogogit function does pretty well on parsing the date/time:


image

**Comment by wxiaoguang** [comment]
@yp05327 after reading the code again, I think we should rewrite the code. If you don't mind, I could propose a new approach soon.

**Comment by yp05327** [comment]
> @yp05327 after reading the code again, I think we should rewrite the code (for both gogit and nogogit). If you don't mind, I could propose a new approach soon.

Never mind. It seems that the old codes haven’t been finished yet. So let’s finish it.

**Pull request closed without merging**